File: TrappedWater.py
#Solution using two pointers
#Time Complexity O(N)
#Space Complexity O(1)

class Solution:
    def trap(self, height: List[int]) -> int:
        left,right = 0,len(height)-1
        left_max = right_max = 0
        amount = 0
        while left<right:
            left_max = max(left_max,height[left])
            right_max = max(right_max,height[right])
            if left_max <= right_max:
                amount += (left_max-height[left])
                left+= 1
            else:
                amount += (right_max-height[right])
                right -= 1
        return amount

File: TrappedWater.cpp
//TC:O(n)
//SC: O(n)
class Solution {
public:
    int trap(vector<int>& height) {
       int n = height.size();
       int result = 0;

       int l = 0, lw = height[0];
       int r = n-1, rw = height[n-1];

       while(l<= r)
       {
           if(lw<rw)
           {
               if(height[l] < lw)
               {
                   result += lw - height[l];
               }else
               {
                   lw = height[l];
               }
               l++;
           }
           else{
               if(height[r] < rw)
               {
                   result += rw - height[r];
               }
               else
               {
                   rw = height[r];
               }
               r--;
           }
       } 
       return result;
    }
};